zvariant_derive/src/value.rs

```diff
@@ -269,34 +297,76 @@ fn impl_enum(
                     #[inline]
                     fn try_from(e: #name) -> #zv::Result<Self> {
                         <#zv::OwnedValue as ::std::convert::TryFrom<_>>::try_from(
-                            <#zv::Value as ::std::convert::From<_>>::from(e as #repr)
+                            <#zv::Value as ::std::convert::From<_>>::from(#into_val)
                         )
                     }
                 }
             },
         ),
     };
 
+    let from_val = if str_enum {
+        quote! {
+            let v: #zv::Str = ::std::convert::TryInto::try_into(value)?;
+
+            ::std::result::Result::Ok(match v.as_str() {
+                #(
+                    #str_values => #name::#variant_names,
+                )*
+                _ => return ::std::result::Result::Err(#zv::Error::IncorrectType),
+            })
+        }
+    } else {
+        quote! {
+            let v: #repr = ::std::convert::TryInto::try_into(value)?;
+
+            ::std::result::Result::Ok(
+                #(
+                    if v == #name::#variant_names as #repr {
+                        #name::#variant_names
+                    } else
+                )* {
+                    return ::std::result::Result::Err(#zv::Error::IncorrectType);
+                }
+            )
+        }
+    };
+
     Ok(quote! {
         impl ::std::convert::TryFrom<#value_type> for #name {
             type Error = #zv::Error;
 
             #[inline]
             fn try_from(value: #value_type) -> #zv::Result<Self> {
-                let v: #repr = ::std::convert::TryInto::try_into(value)?;
-
-                ::std::result::Result::Ok(
-                    #(
-                        if v == #name::#variant_names as #repr {
-                            #name::#variant_names
-                        } else
-                    )* {
-                        return ::std::result::Result::Err(#zv::Error::IncorrectType);
-                    }
-                )
+                #from_val
             }
         }
 
         #into_value
     })
 }
+
+fn enum_name_for_variant(
```

Comment:
`SerializeDict` and `DeserializeDict` already support these renaming attributes so I'm sure there is existing code you can use.

Reply:
Of course, but currently `dict_name_for_field` expect `syn::Field` as an argument, not something like `Literal` or `String`. I would like to avoid modifying existing code which is not related to this PR.

Reply:
It **is** related to this PR since there is no good reason to duplicate the code. AFAIK, it's a pretty trivial change so why not just do it? It will need to be renamed and moved to utils mod btw.

Reply:
Ok